lib/SIL/Utils/Dominance.cpp

```diff
@@ -142,20 +142,26 @@ void PostDominanceInfo::verify() const {
   }
 }
 
-void
-swift::computeDominanceFrontier(SILBasicBlock *root,
-                                DominanceInfo *domTree,
-                                SmallVectorImpl<SILBasicBlock *> &frontier) {
+void swift::computeBlockBoundary(SILBasicBlock *root, DominanceInfo *domTree,
+                                 SmallVectorImpl<SILBasicBlock *> &boundary) {
   auto *function = root->getParent();
   assert(function->hasOwnership());
-  assert(frontier.empty());
+  assert(boundary.empty());
 
   DominanceOrder domOrder(root, domTree);
   while (SILBasicBlock *block = domOrder.getNext()) {
     DominanceInfoNode *domNode = domTree->getNode(block);
     if (domNode->isLeaf()) {
-      frontier.push_back(block);
-      continue;
+      if (block->getNumSuccessors() == 0) {
+        boundary.push_back(block);
+        continue;
+      }
+      for (auto &succ : block->getSuccessors()) {
```

Comment:
The precondition says "OSSA". In that case, a leaf block can only have zero or one successor. If a block has multiple successors, then they will all be dominated.

This might make it a little easier to reason about
```
  if (!domNode->isLeaf()) {
    boundary.push_back(block);
    continue;
  }
  // handle leaf cases...
```

I may be the one who mislead you, but a successor of a leaf block could have multiple predecessors, all of which are dominated. Do you want to just directly do a dominance check on the successor?

Reply:
I thought if the successor of a leaf block had multiple predecessors all of which dominated it, then the leaf block we are looking at won't be a "leaf" ?

Reply:
You are right, I fixed it.